### PR TITLE
Add option to ignore pods issued from a job

### PR DESCRIPTION
**Pull request by fredgate**
Remove annotation from pods not running anymore.

### CODE REVIEW

**Comment by fredgate** [comment]
Fix #6

**Review by duyanghao — COMMENTED** [review]
Some comments for your PR as below.

File: cmd/controller/velerovolume/config/config.go [hunk 1]
```diff
@@ -12,6 +12,7 @@ type VeleroVolumeCfg struct {
 	ExcludeNamespaces  string `yaml:"excludeNamespaces,omitempty"`
 	IncludeVolumeTypes string `yaml:"includeVolumeTypes,omitempty"`
 	ExcludeVolumeTypes string `yaml:"excludeVolumeTypes,omitempty"`
+	ExcludeJobs        bool   `yaml:"excludeJobs"`
```
Comment (duyanghao) [inline cmd/controller/velerovolume/config/config.go]: ExcludeJobs should be list of excluding job names

File: cmd/controller/velerovolume/controller.go [hunk 2]
```diff
@@ -209,6 +209,29 @@ func (c *Controller) syncHandler(key string) error {
 		return err
 	}
 
+	// If pod is pending, we ignore it for moment
```
Comment (duyanghao) [inline cmd/controller/velerovolume/controller.go]: A better way is to add all these filter logics into one function called `checkFilterOptions`

File: cmd/controller/velerovolume/controller.go [hunk 3]
```diff
@@ -209,6 +209,29 @@ func (c *Controller) syncHandler(key string) error {
...
+	if pod.Status.Phase != corev1.PodRunning {
+		err = c.removeBackupAnnotationsFromPod(pod)
+		if err != nil {
+			klog.Errorf("failed to remove velero restic backup annotation to pod: '%s/%s', error: %s", pod.Namespace, pod.Name, err.Error())
+		}
+		return err
+	}
+
+	// Drop pods controlled by a job
+	if c.cfg.ExcludeJobs {
```
Comment (duyanghao) [inline cmd/controller/velerovolume/controller.go]: Add into `checkFilterOptions` as suggested above

File: cmd/controller/velerovolume/controller.go [hunk 4]
```diff
@@ -209,6 +209,29 @@ func (c *Controller) syncHandler(key string) error {
...
+		for _, owner := range pod.OwnerReferences {
+			if owner.Kind == "Job" {
+				klog.V(4).Infof("drop pod: '%s/%s' as it's controlled by a job", pod.Namespace, pod.Name)
+				return nil
+			}
+		}
+	}
+
 	// Drop pods that don't meet namespace requirements
 	if c.cfg.IncludeNamespaces != "" {
```
Comment (duyanghao) [inline cmd/controller/velerovolume/controller.go:260]: Also this one

File: cmd/controller/velerovolume/controller.go [hunk 5]
```diff
@@ -284,6 +307,24 @@ func (c *Controller) addBackupAnnotationsToPod(pod *corev1.Pod) error {
 	return nil
 }
 
+// removeBackupAnnotationsFromPod removes relevant backup annotation from pod.
+func (c *Controller) removeBackupAnnotationsFromPod(pod *corev1.Pod) error {
+	if pod.Annotations != nil {
+		if _, ok := pod.Annotations[constants.VELERO_BACKUP_ANNOTATION_KEY]; ok {
+			// Remove annotation as it is present
+			podCopy := pod.DeepCopy()
```
Comment (duyanghao) [inline cmd/controller/velerovolume/controller.go:336]:
```
if _, exist := podCopy.Annotations[constants.VELERO_BACKUP_ANNOTATION_KEY]; !exist {
	return nil
}
```

File: cmd/controller/velerovolume/controller.go [hunk 6]
```diff
@@ -284,6 +307,24 @@ func (c *Controller) addBackupAnnotationsToPod(pod *corev1.Pod) error {
...
+			delete(podCopy.Annotations, constants.VELERO_BACKUP_ANNOTATION_KEY)
+
+			// Update pod
+			_, err := c.kubeclientset.CoreV1().Pods(pod.Namespace).Update(context.TODO(), podCopy, metav1.UpdateOptions{})
+			if err != nil {
+				return err
+			}
+			klog.V(4).Infof("remove velero restic backup annotation: '%s' to pod '%s/%s' successfully", constants.VELERO_BACKUP_ANNOTATION_KEY, pod.Namespace, pod.Name)
```
Comment (duyanghao) [inline cmd/controller/velerovolume/controller.go]: `klog.V(4).Infof("remove velero restic backup annotation: '%s' from pod '%s/%s' successfully", constants.VELERO_BACKUP_ANNOTATION_KEY, pod.Namespace, pod.Name)`

File: cmd/controller/velerovolume/controller.go (hunk 4 above)
Comment (fredgate) [reply cmd/controller/velerovolume/controller.go:260]: Do you want to add the four checks in this function : pending, running, jobs, namespaces ?
So should it remove the annotation if the pod is not running ? (if this is the case, it does more than check).
So what is the signature of the function : a boolean indicating if the annotation should be added to the pod ?
Or may be the function should check only 3 criteria : pending, jobs, namespaces. And add or remove of the annotation is kept in the `syncHandler` function.

File: cmd/controller/velerovolume/config/config.go (hunk 1 above)
Comment (fredgate) [reply cmd/controller/velerovolume/config/config.go]: I don't think that it is a good idea.
First I would like to be able to ignore all jobs, ignoring their names and count (some jobs can be later deployed in the cluster and I would like that the controller continue to ignore the new jobs without reconfiguring it all the times). Jobs are often used to update volumes used by another pods, so backup can be done from these pods.
Then jobs can be created by cronjob, so we can not know their name, as each time a new job is scheduled, a new name is assigned to it.

File: cmd/controller/velerovolume/controller.go (hunk 6 above)
Comment (fredgate) [reply cmd/controller/velerovolume/controller.go]: Done.
I also reported for log when an error occurs during removal.

File: cmd/controller/velerovolume/controller.go (hunk 4 above)
Comment (fredgate) [reply cmd/controller/velerovolume/controller.go:260]: I added the `checkFilterOptions` function that checks if the pod satisfies the requirements (not pending, jobs, namspaces).
Then if the pod is accepted, the `syncHandler` function updates the pod the right way according to its status, by adding or removing the annotation.

**Comment by duyanghao** [comment]
@fredgate As I pointed out earlier in the comments:
Consider that user maybe not want to drop backup for all jobs, so again ExcludeJobs should be list of excluding job names, and this filed can support cronjob through regex syntax or achieve a simple name-prefix like below:

```
* - exclude all pods owned by job
test,test1 - exclude pods owned by job.name=test or job.name=test1
test*,test1* - exclude pods owned by job.name begins with test or test1(cronjob Scenario)
others...
``` 

I will merge this PR after this problem is resolved.

**Comment by fredgate** [comment]
Ok for using a glob pattern allowing `*` character. In this case, we can ignore all jobs by setting `ExcludeJobs` to `*`.

**Comment by fredgate** [comment]
Changed. Tested on a cluster with `*` and jobs are well skipped.

**Comment by duyanghao** [comment]
@fredgate LGTM, Many thanks for your great work.

**Pull request merged**